File: Sorts/ShellSort.py
"""
Brian Taylor Vann
github.com/taylor-vann

Description:
- Module to shell sort a list

Requirements:
- None

Methods:
- shell_sort
"""


def shell_sort(a):
    n = len(a)

    if n <= 1:
        return

    spc = n

    while spc > 0:
        for i in range(spc):
            prv = i
            idx = i + spc

            while idx > 0 and a[idx - spc] > a[idx]:
                    a[idx], a[idx - spc] = a[idx - spc], a[idx]
                    idx -= spc
            prv += spc
            idx = prv + spc

        spc //= 2

File: ECMAScript/Sorts/ShellSort.js
/*
Brian Taylor Vann
github.com/taylor-vann

Description:
- Function to shell sort a list

Requirements:
- None

Methods:
- shellSort()
*/

function shellSort(a) {
  l = a.length;

  if(l < 2) {
    return a;
  }

  // make space half of array length
  spc = Math.floor(l / 2);

  // while space is greater than 1
  while(spc > 0) {
    // run through everl element between space and array length
    for(var j = spc; j < l; j++) {
      // modified insertion sort using space
      while(j > -1 && a[j - spc] > a[j]) {
        var tmp = a[j - spc];
        a[j - spc] = a[j];
        a[j] = tmp;

        j -= spc;
      }
    }

    spc = Math.floor(spc / 2);
  }

  return a;
}

console.log(shellSort([]));
console.log(shellSort([1]));
console.log(shellSort([5, 2, 6, 8, 9, 2]));
console.log(shellSort([2, 4, 5]));
console.log(shellSort([7, 3, 6, 7]));
console.log(shellSort([2, 4, 1, 9, 3]));
